fe/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java

```diff
@@ -0,0 +1,206 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This class is used to periodically add or drop partition on an olapTable which specify dynamic partition properties
+ * Config.dynamic_partition_enable determine whether this feature is enable, Config.dynamic_partition_check_interval_seconds
+ * determine how often the task is performed
+ */
+public class DynamicPartitionScheduler extends MasterDaemon {
+    private static final Logger LOG = LogManager.getLogger(DynamicPartitionScheduler.class);
+
+    private final String defaultValue = "N/A";
+    public String lastSchedulerTime = defaultValue;
+    public String lastUpdateTime = defaultValue;
+    public State dynamicPartitionState = State.NORMAL;
+    public String msg = defaultValue;
+
+    public enum State {
+        NORMAL,
+        ERROR
+    }
+
+    private Set<Pair<Long, Long>> dynamicPartitionTableInfo = new HashSet<>();
+    private boolean initialize;
+
+    public DynamicPartitionScheduler(String name, long intervalMs) {
+        super(name, intervalMs);
+        this.initialize = false;
+    }
+    public synchronized void registerDynamicPartitionTable(Long dbId, Long tableId) {
+        dynamicPartitionTableInfo.add(new Pair<>(dbId, tableId));
+    }
+
+    public synchronized void removeDynamicPartitionTable(Long dbId, Long tableId) {
+        dynamicPartitionTableInfo.remove(new Pair<>(dbId, tableId));
+    }
+
+    private void dynamicAddPartition() {
+        for (Pair<Long, Long> tableInfo : dynamicPartitionTableInfo) {
+            Long dbId = tableInfo.first;
+            Long tableId = tableInfo.second;
+            Database db = Catalog.getInstance().getDb(dbId);
+            if (db == null) {
+                removeDynamicPartitionTable(dbId, tableId);
+                continue;
+            }
+            db.readLock();
+            try {
+                Table table = db.getTable(tableId);
+                if (table == null ||
+                        !Boolean.parseBoolean(((OlapTable) table).getTableProperty().getDynamicPartitionProperty().getEnable())) {
+                    removeDynamicPartitionTable(dbId, tableId);
+                    continue;
+                }
+            } finally {
+                db.readUnlock();
+            }
+
+            // Determine the partition column type
+            // if column type is Date, format partition name as yyyyMMdd
+            // if column type is DateTime, format partition name as yyyyMMddHHssmm
+            OlapTable olapTable = (OlapTable) db.getTable(tableId);
+            RangePartitionInfo rangePartitionInfo = (RangePartitionInfo) olapTable.getPartitionInfo();
+            Column partitionColumn = rangePartitionInfo.getPartitionColumns().get(0);
+            String partitionFormat = DynamicPartitionUtil.getPartitionFormat(partitionColumn);
+
+            Calendar calendar = Calendar.getInstance();
+            TableProperty tableProperty = olapTable.getTableProperty();
```

Comment:
better to get getDynamicPartitionProperty() here to avoid get it each time.

Reply:
Done